CIP Cardano Multichain Router b02eb0e3a2bf408a85bdbe6a4ce5603b.md

```diff
@@ -0,0 +1,395 @@
+
+ Multichain protocol use transactions chaining on Cardano to improve the concurrency. 
+
+### Burning the wrapped token on Cardano
+
+If the balance of the swapin token on SMPC address is enough, the SMPC nodes will transfer the the needed amount.
+
+If the balance is not enough, the SMPC nodes will mint the lacking amount.
+
+When users bridged wrapped token on Cardano from Cardano to Ethereum, the wrapped tokens will be locked on the MPC address. A independent thread periodically burn the tokens locked on MPC address every week. 
+
+### Configuraton
+
+All the configuration is stored on a onchian smart contract.
+
+```json
+// key is chainID
+mapping(uint256 => ChainConfig) _chainConfig;
+
+// key is tokenID,chainID
+mapping(string => mapping(uint256 => TokenConfig)) _tokenConfig; 
+
+// key is tokenID,srcChainID,dstChainID
+mapping(string => mapping(uint256 => mapping(uint256 => SwapConfig))) _swapConfig; 
+ 
+// key is tokenID,srcChainID,dstChainID
+mapping(string => mapping(uint256 => mapping(uint256 => FeeConfig))) _feeConfig;
+
+struct ChainConfig {
+        string BlockChain;
+        string RouterContract;
+        uint64 Confirmations;
+        uint64 InitialHeight;
+        string Extra;
+    }
+
+struct TokenConfig {
+        uint8 Decimals;
+        string ContractAddress;
+        uint256 ContractVersion;
+        string RouterContract;
+        string Extra;
+    }
+
+struct SwapConfig {
+        uint256 MaximumSwap;
+        uint256 MinimumSwap;
+        uint256 BigValueThreshold;
+    }
+
+struct FeeConfig {
+        uint256 MaximumSwapFee;
+        uint256 MinimumSwapFee;
+        uint256 SwapFeeRatePerMillion;
+    }
+
+```
+
+# Codebase Guide
+
+## Major Module
+
+- [Transaction Register](https://github.com/anyswap/CrossChain-Router/blob/75f2953e496ca02b3d92b8eeb31df54b0f35ae7a/tokens/cardano/register.go#L23) RegisterSwap
+- [Transaction Build](https://github.com/anyswap/CrossChain-Router/blob/75f2953e496ca02b3d92b8eeb31df54b0f35ae7a/tokens/cardano/buildtx.go#L20) BuildRawTransaction
+- [MPC Sign](https://github.com/anyswap/CrossChain-Router/blob/75f2953e496ca02b3d92b8eeb31df54b0f35ae7a/tokens/cardano/signtx.go#L23) MPCSignTransaction
+
+## Functions
+
+- [Query Transaction](https://github.com/anyswap/CrossChain-Router/blob/75f2953e496ca02b3d92b8eeb31df54b0f35ae7a/tokens/cardano/verifytx.go#L96) getTxOutputs
+- [Fees/Decimal Process](https://github.com/anyswap/CrossChain-Router/blob/75f2953e496ca02b3d92b8eeb31df54b0f35ae7a/tokens/base.go#L232) CheckTokenSwapValue
+- [Selecting UTXO](https://github.com/anyswap/CrossChain-Router/blob/75f2953e496ca02b3d92b8eeb31df54b0f35ae7a/tokens/cardano/buildtx.go#L292) BuildTx
+- [Cardano-CLI](https://github.com/anyswap/CrossChain-Router/blob/75f2953e496ca02b3d92b8eeb31df54b0f35ae7a/tokens/cardano/cardanoCmd.go#L42) ExecCmd
+
+# Flow Charts
+
+[How offchain cached utxo work.pdf](CIP%20Cardano%20Multichain%20Router%20b02eb0e3a2bf408a85bdbe6a4ce5603b/How_offchain_cached_utxo_work.pdf)
```

Comment:
Additionally in the flow diagram. Do the MPC share a common utxo cache? What would happen if there are 2 swapins running at the same time and the MPC would sign 2 transactions using the same input UTxOs? Will thinks be retried?